### ISSUE TITLE
Kudu ZipDeploy through Azure CLI fails with 'cannot access the file 'D:\\home\\site\\deployments\\active' ... is being used by another process.'

### ISSUE THREAD

**Issue by JiriKovar** [projectkudu/kudu#3083]
Hi, 

During the past couple of weeks, we've started getting lots of errors 'The process cannot access the file 'D:\\home\\site\\deployments\\active' because it is being used by another process.' when deploying to web app via Kudu ZipDeploy run through Azure CLI ('az webapp deployment source config-zip' command). The problem seems to be that Azure CLI is polling for deployment completion and these requests can clash with Kudu DeploymentManager.

Here is the error from trace log of the deployment:
`<step title="Error occurred" date="2019-11-18T15:20:22.537" type="error" text="The process cannot access the file &apos;D:\home\site\deployments\active&apos; because it is being used by another process." stackTrace="   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize)
   at System.IO.File.OpenFile(String path, FileAccess access, SafeFileHandle&amp; handle)
   at System.IO.File.SetLastWriteTimeUtc(String path, DateTime lastWriteTimeUtc)
   at System.IO.Abstractions.FileWrapper.SetLastWriteTimeUtc(String path, DateTime lastWriteTimeUtc)
   at Kudu.Core.Infrastructure.FileSystemHelpers.SetLastWriteTimeUtc(String path, DateTime lastWriteTimeUtc) in C:\Kudu Files\Private\src\master\Kudu.Core\Infrastructure\FileSystemHelpers.cs:line 259
   at Kudu.Contracts.Infrastructure.LockExtensions.TryLockOperation(IOperationLock lockObj, Action operation, String operationName, TimeSpan timeout) in C:\Kudu Files\Private\src\master\Kudu.Contracts\Infrastructure\LockExtensions.cs:line 35
   at Kudu.Contracts.Infrastructure.LockExtensions.LockOperation(IOperationLock lockObj, Action operation, String operationName, TimeSpan timeout) in C:\Kudu Files\Private\src\master\Kudu.Contracts\Infrastructure\LockExtensions.cs:line 46
   at Kudu.Core.Deployment.DeploymentStatusFile.Save() in C:\Kudu Files\Private\src\master\Kudu.Core\Deployment\DeploymentStatusFile.cs:line 197
   at Kudu.Core.Deployment.DeploymentManager.&lt;Build&gt;d__37.MoveNext() in C:\Kudu Files\Private\src\master\Kudu.Core\Deployment\DeploymentManager.cs:line 589" />
    <step title="WebHooksManager.PublishEventAsync: PostDeployment" date="2019-11-18T15:20:22.552" />
  </step>
</step>
`

And here is the trace log of the last request:

`<step title="Incoming Request" date="2019-11-18T15:20:22.089" instance="ea3dc5" url="/api/deployments/latest" method="GET" type="request" pid="1696,2,50" Connection="Keep-Alive" Content-Type="application/octet-stream" Accept="*/*" Accept-Encoding="gzip, deflate" Authorization="Bas..." Host="draft-resource-intensive-webjobs-aueast.scm.azurewebsites.net" User-Agent="AZURECLI/2.0.74" X-WAWS-Unencoded-URL="/api/deployments/latest" CLIENT-IP="85.93.116.83:64845" X-ARR-LOG-ID="eeb8506b-986c-464e-b934-58e4c85d7de6" X-SITE-DEPLOYMENT-ID="draft-resource-intensive-webjobs-aueast" WAS-DEFAULT-HOSTNAME="draft-resource-intensive-webjobs-aueast.scm.azurewebsites.net" X-Forwarded-Proto="https" X-AppService-Proto="https" >
  <step title="Cleanup Xml Logs" date="2019-11-18T15:20:22.099" />
  <step title="DeploymentService.GetResult" date="2019-11-18T15:20:22.099" >
    <step title="DeploymentService.GetLatestDeployment" date="2019-11-18T15:20:22.115" >
      <step title="DeploymentManager.GetResults" date="2019-11-18T15:20:22.115" />
      <step title="Deployment 179fc72234594f28a7af1e227e4b2ecc is Building" date="2019-11-18T15:20:22.802" />
    </step>
  </step>
  <step title="Outgoing response" date="2019-11-18T15:20:22.818" type="response" statusCode="202" statusText="Accepted" Server="Microsoft-IIS/10.0" x-ms-request-id="eeb8506b-986c-464e-b934-58e4c85d7de6" Cache-Control="private" X-AspNet-Version="4.0.30319" Location="https://draft-resource-intensive-webjobs-aueast.scm.azurewebsites.net/api/deployments/latest" Content-Type="application/json; charset=utf-8" />
</step>`

Please notice the time "15:20:22", because that is when the request is running and when the error occurs. Also please know that there are no other requests or utilities running on that web app at that moment on our end (this particular failure occurred on web app that is running only a scheduled webjobs that weren't even running at that moment, however it happens across all of our web apps randomly).

Is there anything we can do about that apart of trying again? Would it be possible to prevent the request for latest deployment from holding the file?

Thank you very much!

**Comment by suwatch** [comment]
We do have lock in place (could be rare race).   You mentioned "lots" of errors.   Per our log, it happened to your webapp once in the last 14 days.   Around 3% of all deployments ran into this (so not wide spread).  We will prioritize the fix accordingly.

**Comment by JiriKovar** [comment]
Thanks for a quick reply and please let me explain what I meant by "lots" of errors. 

According to our metric (which are based on Azure Cost Analysis and Azure DevOps), we are running more then 10000 zipdeploys a month. 2/3 of them are on our temporary environments for tests that are deleted after completion. That still makes up for a "lots" (11) failures on daily basis even if only 3% of those fail and our developers are asking questions. Most of our deployments consists of deployments across 20+ services and its currently more probable to meet a failure then succeed without any problem (but this issue is not the only one causing that).

The biggest issue is that we rarely saw any kind of deployment problem until a few weeks ago. The number of errors rose so much that we had to look into the problems as and found at least two of them - this is one of them. Another one is the problem with autoswap, where it looks like Azure CLI is unable to deal with the autoswap which leads to various errors. I have created a separate issue in their repository as I guessed that it should be handled on their side: https://github.com/Azure/azure-cli/issues/11343

**Comment by suwatch** [comment]
We really appreciate you reporting issues.  For the past 2 weeks (11/10/2019 - today), the App Service has gone into upgrade.  During such operations, often times file system could go into read/only mode which could cause deployment to fail (unable to write files).   At the end of the day, we need to see the percentage of those to help justify/prioritize the improvement.   Taking auto-swap out of the equation, it would be great to know what the error rate looks like for your usages.

**Comment by JiriKovar** [comment]
Unfortunately I won't be able to provide you with those statistics as it's hard to distinguish the cause of the errors. The Azure CLI is not very helpful in reporting the details of the failure (and I understand why) and we need to look into Kudu trace logs and "make an educated guess" (like in this case - there was no proof to be found that the file is really being held by the request from Azure CLI). It might be possible to create some kind of heuristic that would try to analyze the trace logs and make the educated guesses in our stead, but it would require quite a lot of work on our end and I'd guess that fixing the issue you already know of should be easier. 

I can think of multiple solutions out of the top of my head - for example a extremely simple retry logic - if it happens only in 3% of the cases, then retrying the operation 3 times with 0.5s delay should be more than sufficient - the processing of the request takes around 300ms and Azure CLI is polling with 2s delay (which is honestly more then necessary) and delaying the deployment by 1.5s means nothing compared to unnecessary failure. I know it's not a perfect solution, but you can still log the retries to keep the statistics and can have an issue for a better solution in the backlog.

Are you seriously trying to tell me that its better to explain and wait for statistics and prioritize and waste the time of your customers to go through this kind of investigation, compared to implement this simple fix? Please find a decent product owner/manager and make your processes more lightweight.

**Comment by suwatch** [comment]
I apologize if I made it sound that way.  We do appreciate you are using our products - especially reporting to make our product better.  My point is one (decent or not) would always need to prioritize works - often times stat and how critical is one criteria.   Believe me, everyone wants different issues addressed.  If we have more resources than works, we don't have to discuss this.  Most times more backlogs than people to handle - regardless of what the fix would be as it may lead to another issue and so on.

Concerning the fix, yes at first I thought it would be that simple.   Looking at our codes (which we already had to spend times away from other issues), we do have lock and synchronization in place.  Hence, we like to understand and spend more time in case we have a more subtle issue.  Yes we could just patch with retry as you suggested if we can't find any better solution.

My simple ask is, given you are the only one reported, how often this happens to get the sense if we need to drop other things to help with the issue (more investigation and eventually fix it).

**Comment by JiriKovar** [comment]
First of all, I'm sorry for being too frank about my dissatisfaction with how this problem unfolds. I recognize you are doing a great job answering me. It was fast, you provided all the information right away and you were very honest. You have my gratitude.

I understand your need to do things properly and I understand that you need to prioritize. However I don't understand 2 things:

1. What would you break by retry logic in this scenario? The release is about to fail. What worse can happen if you simply try to get an access to this file again? I have to admit that implying that the fix might lead to another problem in this case makes me a bit worried (particularly implying that something worse than deployment failure might happen - otherwise there would be no point in mentioning it and you could give it a shot and we would be able to give you feedback on whether or not it helped).
2. I don't understand why would you spend more time arguing with me than would be required to patch this up and then keep prioritizing on proper fix/refactoring as I already mentioned. This way you save time by not arguing with me, not spending time to understand your code, preventing another people from reporting it again and going through this cycle again and thus preventing another waste of time not only on your end, and will fix 3% of your deployments you already know are failing and can postpone the expensive fix until the time comes for one reason or another. That's a win-win situation, right? If not, could you explain the downsides? I can even prepare you a pull request with the retry logic if you'd want, but I would want to avoid those downsides. 

I want to ask 1 more thing: Are you the one to make this decision, or would it be better if I tried to reach someone else? If so, could you please direct me to them? I wouldn't want to make you argue with me simply because you can't give me a better answer, but I also can't just let it be and pretend like those 11 releases a day doesn't fail.

FYI I have also created a support ticket for this problem (sorry about that - I first created the ticket, but support wasn't answering, so I tried my luck here and you answered before they even wrote the initial email). I have already explained the situation, however they didn't close the ticket yet, so if you/your manager would prefer to contact me via more private channel, you can reach me either through my the public email of my GitHub account, or the ticket (ID: 119111922002352)

**Comment by suwatch** [comment]
We will address it and the fix should be rolled out worldwide by end of Dec.

**Comment by JiriKovar** [comment]
Thank you very much!

**Comment by acd-hfi** [comment]
Any update on this?  I am seeing this frequently with a Event Hub triggered function when trying to do a zip deploy.  It still happens even when shutting down the function.  After attempts, the entire function locks up and stops consuming messages which is a little concerning to us.

"The process cannot access the file 'D:\home\site\deployments\active' because it is being used by another process."

**Comment by suwatch** [comment]
This is a benign error and, if failed, we will not clean up old log.   It will try to clean up on next deployment.

**Comment by acd-hfi** [comment]
@suwatch Thank you for the update.  This error may be benign, but the results absolutely are not.  As soon as a bad deploy like this happens, the function host appears to crash.  I redeploy roughly a dozen times over the course of 24 hours and still couldn't get the deploy to go through.  I ended up needing to delete the function app and recreating it (I waited roughly 20 minutes, when recreated with the same name the old execution history/etc was still present on it).

**Comment by acd-hfi** [comment]
@suwatch What would you reccomend here?  I had to delete the functions again today to have the deployment go through.  There is something obviously wrong here.  The logs were not helpful for this today.

**Comment by suwatch** [comment]
Please provide UTC time of the deployment and the sitename.   We can cross check with the log.

**Comment by acd-hfi** [comment]
> Please provide UTC time of the deployment and the sitename. We can cross check with the log.

I deleted, renamed, and redeployed the resource last evening.  I was surprised to see the zip deploy fail to a new instance; it seems like instances are reused.  The current sitename is hfiq-iotstream-ci.  Deploys failed at: 6/29/2020 11:05:41 PM UTC, 6/29/2020 11:13:09 PM, 6/29/2020,11:17:11 PM UTC.  Across the 3 logs nothing looks particularly bad.  This isn't the best example b/c of the rename and because these deploys wouldn't have crashed the function host as I was expecting it to be a new instance.

**Comment by acd-hfi** [comment]
Another: 2020-06-30T18:39:09.5369888Z.  A second deploy 'fixed' it several minutes later.  Before the second deploy, however, the functions host crashed.  Also as a note, I am stopping the functions, waiting a minute, and then deploying in AzDO.  This seems to result in less issues, but obviously doesn't resolve the issue.

**Comment by suwatch** [comment]
>The current sitename is hfiq-iotstream-ci. Deploys failed at: 6/29/2020 11:05:41 PM UTC, 6/29/2020 11:13:09 PM, 6/29/2020,11:17:11 PM UTC. 

For the above, the failure was due to conflicting WEBSITE_RUN_FROM_PACKAGE settings.   When the site was created, WEBSITE_RUN_FROM_PACKAGE was set to 1.  However, during DevOp deployment, it was removed.   If your intention is to use  WEBSITE_RUN_FROM_PACKAGE, make sure to enable DevOps RunFromPackage as well.  

We also notice an appSetting  WEBSITE_ENABLE_SYNC_UPDATE_SITE added during DevOps - did you add explicitly?

**Comment by acd-hfi** [comment]
@suwatch WEBSITE_RUN_FROM_PACKAGE is set to 1 during arm deployment directly in the arm file.  Are you saying that the later step to do the zip deploy is removing it?  The AzDO zip deploy is setting more settings. 

So steps are:
1)  Stop functions, wait 1 min
2) Create or update resource via arm that also sets run from package to 1
3) zip deploy to windows host but also set a couple extra secret config values
4) start app.

No, we are not setting WEBSITE_ENABLE_SYNC_UPDATE_SITE anywhere.

**Comment by suwatch** [comment]
> Are you saying that the later step to do the zip deploy is removing it? The AzDO zip deploy is setting more settings.

This is what VSTS update right the site creation (see WEBSITE_RUN_FROM_PACKAGE was removed and WEBSITE_ENABLE_SYNC_UPDATE_SITE added).  You should see the same on Portal AppSettings.

```json
{
  "AzureWebJobsDashboard": "...",
  "AzureWebJobsStorage": "...",
  "WEBSITE_CONTENTAZUREFILECONNECTIONSTRING": "...",
  "WEBSITE_CONTENTSHARE": "...",
  "FUNCTIONS_EXTENSION_VERSION": "...",
  "APPINSIGHTS_INSTRUMENTATIONKEY": "...",
  "ASPNETCORE_ENVIRONMENT": "...",
  "WEBSITE_ENABLE_SYNC_UPDATE_SITE": "...",
}
```

For WEBSITE_RUN_FROM_PACKAGE, AzDO must have an option for you to include this - my apology as I am not familiar with AzDO options and pipeline.   For WEBSITE_ENABLE_SYNC_UPDATE_SITE, you can also set the same for ARM during site creation.

**Comment by acd-hfi** [comment]
@suwatch I am not following.  I tried disabling all but the arm template deployment, and it did not crash the site and the setting is still there (on the last deploys).  The deployment history also shows WEBSITE_RUN_FROM_PACKAGE being set to 1 during the arm deploys.  The AzDO tasks essentaully use the same commands as az commands (I have had these issues with both az and vsts/azdo tasks).  

What are you suggesting with this comment?

**Comment by acd-hfi** [comment]
@suwatch Well huh.  The last deploy that ONLY included the ARM template crashed the functions.  WEBSITE_RUN_FROM_PACKAGE is still 1, however.

**Comment by acd-hfi** [comment]
@suwatch Are we saying that ARM deployments for Event hub functions are not safe?

**Comment by suwatch** [comment]
I am not sure how we go from this to imply whether this is safe or not.  First, like any investigation, we need to know what was the issue with your deployment while others did not run into the same.  The best way to approach this is to pick case by case and try to explain them.  Hopefully, it will show common patterns of an issue and solution.   

> The current sitename is hfiq-iotstream-ci. Deploys failed at: 6/29/2020 11:05:41 PM UTC, 6/29/2020 11:13:09 PM, 6/29/2020,11:17:11 PM UTC.

We picked the above cases.  What we found was the inconsistent appSettings.

@2020-06-29 23:06:48 UTC Site was created with WEBSITE_RUN_FROM_PACKAGE =1 in ARM template
@2020-06-29 23:08:03 UTC Site was updated to remove that setting - from user-agent, it was done by VSTS / DevOps.
@2020-06-29 23:13:36 UTC Site was updated with WEBSITE_RUN_FROM_PACKAGE =1 in ARM template
@2020-06-29 23:15:02 UTC Site was updated to remove that setting - from user-agent, it was done by VSTS / DevOps.
@2020-06-29 23:17:31 UTC Site was updated with WEBSITE_RUN_FROM_PACKAGE =1 in ARM template
@2020-06-29 23:19:00 UTC Site was updated to remove that setting - from user-agent, it was done by VSTS / DevOps.

The WEBSITE_RUN_FROM_PACKAGE dictates how the deployment and runtime should handle the content.  One should not flip flop the value.   We guessed you did not do it on purpose.   Next was to try to understand why DevOps remove the setting!   Unfortunately, this is not a DevOps expert forum - you will need to consult with DevOps or its doc.   I am sure there must be an option to WEBSITE_RUN_FROM_PACKAGE = 1 and not setting WEBSITE_ENABLE_SYNC_UPDATE_SITE - since others are able to do it successfully.

**Comment by acd-hfi** [comment]
@suwatch I think the deploy at 2020-06-30T22:35:08.3388445Z would be a better example to investigate.  The deploy, in Azure, shows that WEBSITE_RUN_FROM_PACKAGE was being set to 1 directly in the arm.  The current config on the function app after arm deploy has the setting set to 1 as well.  I also can't help but feel, regardless of how many times that setting may be appearing to change on your side, that the fact that the website is offline/stopped during these changes should negate any issue potentially caused by the settings changing.

I had to delete the resource again this morning to get the app started again.  It was down since last evening when I only deployed the arm template using the Azure DevOps arm deployment task.

As of today, I modified the arm deployment to use an az cli command instead of the built in Azure DevOps task.  I will update you on the results.
